### ISSUE TITLE
Is the require condition correct at 1.7.355?

### ISSUE THREAD

**Issue by mysterlune** [mozilla/pdf.js#8146]
Link to PDF file (or attach file here): N/A

Configuration:
- Web browser and its version: Chrome 56.0.2924.8
- Operating system and its version:
- PDF.js version: 1.7.355
- Is an extension: not sure what this means...

Steps to reproduce the problem:
1. [attached a screencast below]
2... or... check out https://github.com/mysterlune/ember-pdfjs
3. update `bower.json` to be `pdfjs-dist: "1.7.355"`
4. `npm install && bower install && ember serve` (maybe have to install ember?)

... when you load up the default main view in the app and launch the debugger, you'll see the issue in the screen cast. if it's an issue in the Ember community side of the thing (e.g. with `loader.js`) i'm happy to raise the issue. i'm just wondering if the conditional in this commit is correct...

What is the expected behavior? (add screenshot)

it appears that the commit referenced below has made the code flow enter a case that is perhaps incorrect (perhaps assuming a server-side context?) where the browser's use of a requireJs-ish package (in my case `loader.js`) is causing execution of this `if(typeof require === 'function')` block.

it seems that the check for `typeof` is not really adequate where consumers may be using `requireJs` or something that will make the lookup fail...

What went wrong? (add screenshot)

https://github.com/mozilla/pdfjs-dist/commit/a88a99517c5ddf8ba8ff6514429a9217e5a59a20#diff-81859d79710fc24b108a056f7a7313deL101

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

Here is a screencast of the goings-on:

https://screencast.com/t/Fkz8EZWzZ

**Comment by Snuffleupagus** [comment]
*For reference:* Note that `mozilla/pdfjs-dist` just contains pre-built versions of the `mozilla/pdf.js` library, hence the relevant change was made in PR #8087 (in response to issue #8084).

**Comment by timvandermeij** [comment]
@yurydelendik Could you comment on this?

**Comment by mysterlune** [comment]
i guess what i'm wondering is whether the conditional establishing the value of `pdfjsLib` should be more like:

````
 if (typeof __pdfjsdev_webpack__ === 'undefined') {
  if (window && window['pdfjs-dist/build/pdf']) {
    pdfjsLib = window['pdfjs-dist/build/pdf'];
  }
  else if (typeof require === 'function') {
    pdfjsLib = require('../build/pdf.js');
  }
  else {
    throw 'Neither `require` nor `window` found';
  }
 }
````

**Comment by yurydelendik** [comment]
@mysterlune The #8087 was introduced with webpack in mind, which analyses JS and "finds" `require('../build/pdf.js')` to bundle this module. Not sure if webpack will skip analysis then it will find `if (typeof window !== 'undefined'` branch. You can submit a PR with the solution you mentioned above, but I would ask to test it with examples/webpack/ test case.

**Comment by mysterlune** [comment]
hi there,

yes, i've run the unit tests for the project and see that:

````
Run 788 tests
All unit tests passed.
unit tests runtime was 26.9 seconds
[13:50:04] Finished 'unittest' after 28 s
````

... next, i built the webpack bundles:

````
13:54 rlune@oakm111430f01:~/gitproj/pdf.js/examples/webpack $ node_modules/.bin/webpack 
Hash: 8a854da8c214c32443b1
Version: webpack 1.12.15
Time: 23259ms
Asset                 Size           Chunks             Chunk Names
main.bundle.js        163 kB           0    [emitted]    main
1.1.bundle.js         603 kB           1    [emitted]  
pdf.worker.bundle.js  605 kB           2, 1 [emitted]    pdf.worker
[0] ./main.js         1.17 kB {0}           [built]
+ 4 hidden modules
````

... then fired up the example webpack as you suggested. in the browser, here's what i see...

https://cl.ly/jVnD

... is this correct? if so, then all seems well with the branch i created.

**Comment by cvlmtg** [comment]
Hi, we're having issues as well on electron.

I remember having the same issues with pdf.js a while ago, where it requires pdf.worker.js, and in fact the code inside it is quite different (starting from line 2418). I think the same code should be used inside pdf_viewer.js as well, to avoid problems.

Thanks in advance!

**Comment by yurydelendik** [comment]
@cvlmtg , actually require() as it's used now is better for electon.

> and in fact the code inside it is quite different (starting from line 2418).

not clear what do you mean by that

**Comment by mysterlune** [comment]
hi @yurydelendik, #8154 is updated, built and tested per your suggestion. PR awaiting review. No hurry, just want to be sure if the code looks right and will be merged.

**Comment by yurydelendik** [comment]
@mysterlune thank you, I'll try to check it shortly. (I want to test an idea if it's possible to add require.amd check instead of window as well)

**Comment by yurydelendik** [comment]
Fixed by #8154

**Comment by cvlmtg** [comment]
> > and in fact the code inside it is quite different (starting from line 2418).

> not clear what do you mean by that

@yurydelendik I mean che code is quite different. in version 1.7.302, pdf_viewer.js is:

```
 if (typeof __pdfjsdev_webpack__ === 'undefined') {
  if (typeof require === 'function') {
   pdfjsLib = require('../build/pdf.js');
  } else {
   pdfjsLib = window['pdfjs-dist/build/pdf'];
  }
 }
```

pdf.js is:

```
if (typeof __pdfjsdev_webpack__ === 'undefined') {
 if (typeof window === 'undefined') {
  isWorkerDisabled = true;
  if (typeof require.ensure === 'undefined') {
   require.ensure = require('node-ensure');
  }
  useRequireEnsure = true;
 } else if (typeof require !== 'undefined' && typeof require.ensure === 'function') {
  useRequireEnsure = true;
 }
 if (typeof requirejs !== 'undefined' && requirejs.toUrl) {
  workerSrc = requirejs.toUrl('pdfjs-dist/build/pdf.worker.js');
 }
 var dynamicLoaderSupported = typeof requirejs !== 'undefined' && requirejs.load;
 fakeWorkerFilesLoader = useRequireEnsure ? function (callback) {
  require.ensure([], function () {
   var worker = require('./pdf.worker.js');
   callback(worker.WorkerMessageHandler);
  });
 } : dynamicLoaderSupported ? function (callback) {
  requirejs(['pdfjs-dist/build/pdf.worker'], function (worker) {
   callback(worker.WorkerMessageHandler);
  });
 } : null;
}
```

anyway we're gonna try the new release as soon as possibile, thanks!
